client/allocrunnerv2/health_hook_test.go

```diff
@@ -0,0 +1,170 @@
+package allocrunnerv2
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/nomad/client/allocrunnerv2/interfaces"
+	"github.com/hashicorp/nomad/client/consul"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/hashicorp/nomad/nomad/mock"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/stretchr/testify/require"
+)
+
+// statically assert health hook implements the expected interfaces
+var _ interfaces.RunnerPrerunHook = (*allocHealthWatcherHook)(nil)
+var _ interfaces.RunnerUpdateHook = (*allocHealthWatcherHook)(nil)
+var _ interfaces.RunnerDestroyHook = (*allocHealthWatcherHook)(nil)
+
+// mockHealthSetter implements healthSetter that stores health internally
+type mockHealthSetter struct {
+	setCalls   int
+	clearCalls int
+	healthy    *bool
+	isDeploy   *bool
+	taskEvents map[string]*structs.TaskEvent
+	mu         sync.Mutex
+}
+
+func (m *mockHealthSetter) SetHealth(healthy, isDeploy bool, taskEvents map[string]*structs.TaskEvent) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.setCalls++
+	m.healthy = &healthy
+	m.isDeploy = &isDeploy
+	m.taskEvents = taskEvents
+}
+
+func (m *mockHealthSetter) ClearHealth() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.clearCalls++
+	m.healthy = nil
+	m.isDeploy = nil
+	m.taskEvents = nil
+}
+
+func TestHealthHook_PrerunDestroy(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+
+	b := cstructs.NewAllocBroadcaster()
+	defer b.Close()
+
+	consul := consul.NewMockConsulServiceClient(t)
+	hs := &mockHealthSetter{}
+
+	h := newAllocHealthWatcherHook(testlog.HCLogger(t), mock.Alloc(), hs, b.Listen(), consul)
+
+	// Assert we implemented the right interfaces
+	prerunh, ok := h.(interfaces.RunnerPrerunHook)
+	require.True(ok)
+	_, ok = h.(interfaces.RunnerUpdateHook)
+	require.True(ok)
+	destroyh, ok := h.(interfaces.RunnerDestroyHook)
+	require.True(ok)
+
+	// Prerun
+	require.NoError(prerunh.Prerun(context.Background()))
```

Comment:
Should we assert behavior around when the context is canceled/deadlined?

Reply:
Great question. This particular hook doesn't use the context and only receives one to satisfy an interface.